trieste/models/gpflux/sampler.py

```diff
@@ -0,0 +1,170 @@
+        if isinstance(layer, GPLayer):
+            function_draws.append(layer.sample())
+        elif isinstance(layer, LatentVariableLayer):
+            function_draws.append(sample_consistent_lv_layer(layer))
+        else:
+            raise NotImplementedError(f"Sampling not implemented for {type(layer)}")
+
+    class ChainedSample(Sample):  # type: ignore[misc]
+        def __call__(self, X: TensorType) -> tf.Tensor:
+            for f in function_draws:
+                X = f(X)
+            return X
+
+    return ChainedSample().__call__
+
+
+class DeepGaussianProcessTrajectorySampler(TrajectorySampler[GPfluxPredictor]):
+    r"""
+    This sampler provides trajectory samples from a :class:`DeepGaussianProcess`\ 's predictive
+    distribution.
+    """
+
+    def __init__(self, model: ProbabilisticModel):
+        """
+        :param model: The model to sample from. Must be a :class:`GPfluxPredictor`
+        :raise ValueError: If model is not a :class:`GPfluxPredictor`
+        """
+        if not isinstance(model, GPfluxPredictor):
+            raise ValueError(
+                f"Model must be a gpflux.interface.GPfluxPredictor, received {type(model)}"
+            )
+
+        super().__init__(model)
+
+        self.model = model.model_gpflux
+
+        if not isinstance(self.model, DeepGP):
+            raise ValueError(
+                f"GPflux model must be a gpflux.models.DeepGP, received {type:self.model}"
+            )
+
+    def get_trajectory(self) -> TrajectoryFunction:
+        """
+        Generate an approximate function draw (trajectory) by using the GPflux sampling
+        functionality. These trajectories are differentiable with respect to the input, so can be
+        used to e.g. find the minima of Thompson samples.
+
+        :return: A trajectory function representing an approximate trajectory from the deep Gaussian
+            process, taking an input of shape `[N, D]` and returning shape `[N, 1]`
+        """
+
+        return sample_dgp(self.model)
+
+
+class DeepGaussianProcessReparamSampler(ReparametrizationSampler[GPfluxPredictor]):
+    r"""
+    This sampler employs the *reparameterization trick* to approximate samples from a
+    :class:`DeepGaussianProcess`\ 's predictive distribution.
+    """
+
+    def __init__(self, sample_size: int, model: ProbabilisticModel):
+        """
+        :param sample_size: The number of samples for each batch of points. Must be positive.
+        :param model: The model to sample from. Must be a :class:`GPfluxPredictor`
+        :raise ValueError (or InvalidArgumentError): If ``sample_size`` is not positive, or if
+            model is not a :class:`GPfluxPredictor`.
+        """
+        if not isinstance(model, GPfluxPredictor):
+            raise ValueError(
+                f"Model must be a gpflux.interface.GPfluxPredictor, received {type(model)}"
+            )
+
+        super().__init__(sample_size, model)
+
+        self.model = model.model_gpflux
```

Comment:
here you dont check if its a DeepGP class?

Reply:
Good catch, will add that!